src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -59,6 +59,7 @@
 #include "distributed/reference_table_utils.h"
 #include "distributed/relation_access_tracking.h"
 #include "distributed/remote_commands.h"
+#include "distributed/resource_lock.h"
```

Comment:
A semi-related issue, which we might maybe skip for now if the user is not going to be hit by this. 

If the cluster is fresh new (or there are only 2 tables in general), then:

-- session 1
```SQL
CREATE TABLE ref_1 (a int);
SELECT create_reference_table('ref_1');

BEGIN;
DROP TABLE ref1;
```

Session 2 gets blocked:
```SQL
BEGIN;
CREATE TABLE ref_2 (a int);
SELECT create_reference_table('ref_2');
```

On the colocated table id lock:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007ff8065a834e libsystem_kernel.dylib`kevent + 10
    frame #1: 0x0000000106193643 postgres`WaitEventSetWaitBlock(set=0x00007fdaf200a4f8, cur_timeout=<unavailable>, occurred_events=0x00007ff7ba08b5c8, nevents=1) at latch.c:1600:7
    frame #2: 0x0000000106192ff1 postgres`WaitEventSetWait(set=0x00007fdaf200a4f8, timeout=<unavailable>, occurred_events=0x00007ff7ba08b5c8, nevents=1, wait_event_info=50331648) at latch.c:1395:8
    frame #3: 0x0000000106192dd8 postgres`WaitLatch(latch=<unavailable>, wakeEvents=33, timeout=0, wait_event_info=50331648) at latch.c:472:6
    frame #4: 0x00000001061b09c1 postgres`ProcSleep(locallock=<unavailable>, lockMethodTable=<unavailable>) at proc.c:1361:11
    frame #5: 0x00000001061a443c postgres`WaitOnLock(locallock=0x00007fdaf400bbc8, owner=0x00007fdaf401a5c0) at lock.c:1858:7
    frame #6: 0x00000001061a358e postgres`LockAcquireExtended(locktag=0x00007ff7ba08b898, lockmode=1, sessionLock=<unavailable>, dontWait=false, reportMemoryError=false, locallockp=<unavailable>) at lock.c:1100:3
    frame #7: 0x00000001061a1b53 postgres`LockRelationOid(relid=<unavailable>, lockmode=1) at lmgr.c:117:8
    frame #8: 0x0000000106e7bb47 citus.so`ColocatedTableId(colocationId=2) at colocation_utils.c:1162:3
    frame #9: 0x0000000106dccc27 citus.so`CreateDistributedTable(relationId=25848, distributionColumnName=0x0000000000000000, distributionMethod='n', shardCount=<unavailable>, shardCountIsStrict=false, colocateWithTableName=0x0000000000000000, viaDeprecatedAPI=<unavailable>) at create_distributed_tab
```

Reply:
yeah, I noticed that as well. If you look close at the isolation tests I left a comment to that extent. To prevent the isolation tests to run into this I create a reference table in the setup for this locking purpose.

It is a fallacy in how Citus doesn't have shardgroups ;) That lock is more of a shardgroup lock to make sure that you have placements that one can colocate with. After 11.1 I hope to actually introduce shardgroups and move placements from shards to shardgroups. Then we can remove the lock on the first table we find in a colocation group.